File: src/main/scala/com/bytetrend/sandbox/scala/algo/LongestPalindromeInString.scala
package com.bytetrend.sandbox.scala

object LongestPalindromeInString {

  import util.control.Breaks._

  def findBiggestPalindromeSubstring(input: String): String = {
    var palindrome = ""
    var tmpStr = ""
    for (i <- 0 until input.length - 1) {
      if (input.charAt(i) == input.charAt(i + 1)) {
        breakable {
          for (
            //The operation will drop the tail of longer sequence either j or k
            (j, k) <- (i to 0 by -1).zip(i + 1 until input.length)
          ) {
            if (input.charAt(j) == input.charAt(k)) {
              tmpStr = input.substring(j, k + 1)
              if (tmpStr.length > palindrome.length)
                palindrome = tmpStr
            } else break
          }
        }
        if (i > 0 && input.charAt(i - 1) == input.charAt(i + 1)) {
          for (
            (j, k) <- (i - 1 to 0 by -1).zip(i + 1 until input.length)
          ) {
            breakable {
              if (input.charAt(j) == input.charAt(k)) {
                tmpStr = input.substring(j, k + 1)
                if (tmpStr.length > palindrome.length)
                  palindrome = tmpStr
              } else break
            }
          }
        }
      }
    }
    palindrome
  }

  def main(args: Array[String]) = {
    val input = "aaabbaaaccdeqjncsdddmmmkkkmmmddd";
    val answer = "dddmmmkkkmmmddd";
    val resp = findBiggestPalindromeSubstring(input)
    println(s"$resp and expected answer $answer match=${answer.equals(resp)}")
  }

}

File: src/main/java/com/bytetrend/sandbox/java/algo/LongestPalindromeInString.java
package com.bytetrend.sandbox.java.algo;


public class LongestPalindromeInString {
    public static String findBiggestPalindromeSubstring(String input) {
        String tmp;
        String palindrome = "";
        for (int i = 0; i < input.length() - 1; i++) {
            // Even palindrome when two consecutive characters are the same.
            if (input.charAt(i) == input.charAt(i + 1)) {
                for (int j = i, k = i + 1; j >= 0 && k < input.length(); j--, k++) {
                    if (input.charAt(j) == input.charAt(k)) {
                        tmp = input.substring(j, k + 1);
                        if (tmp.length() > palindrome.length())
                            palindrome = tmp;
                    } else
                        break;
                }
            }
            // Odd palindrome when two characters around the current one are the same.
            if (i > 0 && input.charAt(i - 1) == input.charAt(i + 1)) {
                for (int j = i - 1, k = i + 1; j >= 0 && k < input.length(); j--, k++) {
                    if (input.charAt(j) == input.charAt(k)) {
                        tmp = input.substring(j, k + 1);
                        if (tmp.length() > palindrome.length())
                            palindrome = tmp;
                    } else
                        break;
                }
            }
        }
        return palindrome;
    }

    public static String longestPal2(String a) {
        // create new array to hold results
        int[][] arr = new int[a.length() + 1][a.length() + 1];
        // reverse   String a
        String b = new StringBuffer(a).reverse().toString();

        // complete the Longest common subsequence array
        for (int i = 0; i < a.length(); i++) {
            for (int j = 0; j < b.length(); j++) {
                char achar = a.charAt(i);
                char bchar = b.charAt(j);
                if (achar == bchar) {
                    arr[i + 1][j + 1] = arr[i][j] + 1;
                } else {
                    arr[i + 1][j + 1] = Math.max(arr[i + 1][j], arr[i][j + 1]);
                }
            }
        }
        String str = "";
        int x = a.length();
        int y = b.length();
        //trace back form bottom of LCS array
        while (x > 0 && y > 0) {
            if (arr[x][y] == arr[x - 1][y]) {
                x--;
            } else if (arr[x][y] == arr[x][y - 1]) {
                y--;
            } else {
                if (a.charAt(x - 1) == b.charAt(y - 1)) {
                    str += a.charAt(x - 1);
                    x--;
                    y--;
                }
            }
        }
        // return the largest palindrome
        return str;
    }

    public static void main(String[] args) {
        String input = "aaabbaaaccdeqjncsdddmmmkkkmmmddd";
        String answer = "dddmmmkkkmmmddd";
        LongestPalindromeInString pal = new LongestPalindromeInString();
        String resp = pal.longestPal2(input);
        System.out.println(resp.equals(answer));
        String resp2 = pal.findBiggestPalindromeSubstring(input);
        System.out.println(resp2.equals(answer));
    }
}
